#   eBay Kleinanzeigen Used Cars Sales

we'll work with a dataset of used cars from `eBay Kleinanzeigen`, a classifieds section of the German eBay website.The aim of this project is to clean the data and analyze the included used car listings.

This dataset includes the following columns.

1. `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
2. `name` - Name of the car.
3. `seller` - Whether the seller is private or a dealer.
4. `offerType` - The type of listing
5. `price` - The price on the ad to sell the car.
6. `abtest` - Whether the listing is included in an A/B test.
7. `vehicleType` - The vehicle Type.
8. `yearOfRegistration` - The year in which the car was first registered.
9. `gearbox` - The transmission type.
10. `powerPS` - The power of the car in PS.
11. `model` - The car model name.
12. `odometer` - How many kilometers the car has driven.
13. `monthOfRegistration` - The month in which the car was first registered.
14. `fuelType` - What type of fuel the car uses.
15. `brand` - The brand of the car.
16. `notRepairedDamage` - If the car has a damage which is not yet repaired.
17. `dateCreated` - The date on which the eBay listing was created.
18. `nrOfPictures` - The number of pictures in the ad.
19. `postalCode` - The postal code for the location of the vehicle.
20. `lastSeenOnline` - When the crawler saw this ad last online.

In [1]:
# importing the necessary libraries
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

In [2]:
# reading the csv file
autos = pd.read_csv('data/autos.csv', encoding = 'Latin-1')

### Checking the Dataset

In [3]:
# first five rows
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

- Upon initial analysis, the dataset reveals intriguing insights. Comprising 5,000 rows, it consists of `5 integer`-type columns and `15 object`-type columns. The integer columns, representing monthOfRegistration, postalCode etc.. and object columns, storing name, seller, price vehicleType etc... 

- one interesting observation is **price** column is treated as a object type, so i have to carefully check each columns and its data types. However, the presence of null values in certain columns necessitates a careful cleaning process.

- Additionally, a check for duplicate entries is on the agenda. Identifying and handling duplicates is crucial to maintain the accuracy of subsequent analyses.

In [5]:
# print column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

### Column Renaming

The first step in the data cleaning process is renaming the column names, because the column headers were in *camelCase* instead of the conventional python *snake_case*, we will rename the names of the columns to conform to python conventions. To do this, using the method `autos.rename()`.

In [6]:
autos.rename(columns = {'dateCrawled':'data_crawled', 'offerType': 'offer_type',
                       'vehicleType':'vehicle_type', 'yearOfRegistration':
                       'registration_year','monthOfRegistration': 'registration_month',
                       'notRepairedDamage':'unrepaired_damage','fuelType':'fuel_type',
                       'dateCreated':'ad_created','postalCode':'postal_code',
                       'lastSeen':'last_seen','nrOfPictures':'no_of_pictures',
                       'odometer':'odometer_km'}, inplace = True)

Check if the column names have been changed properly.

In [7]:
autos.columns

Index(['data_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [8]:
autos.describe(include = 'all')

data_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  no_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

- The columns `seller`, `offer_type`, `abtest`, `gearbox`, and `unrepaired_damage` contains just 2 unique values. so i will further check each columns and decide whetjer to keep that column or drop that column, because if all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.

- The columns `price` and `odometer` are in object datatype and both of these columns can be converted from string to integer values. 

- The column `registration_year` need to be investigated beacuse minimum value present in the column is `1000` and the maximum value showed is `9999`, which means i have to investigate that column.

In [9]:
def check_unique_values(data, column_name_list):
    for column in column_name_list:
        unique_values = data[column].value_counts()
        print(f'Unique values present in the column {column}:\n{unique_values}')

In [10]:
column_name_list = ['seller', 'offer_type', 'abtest', 'gearbox', 'unrepaired_damage']
check_unique_values(autos, column_name_list)

Unique values present in the column seller:
privat        49999
gewerblich        1
Name: seller, dtype: int64
Unique values present in the column offer_type:
Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64
Unique values present in the column abtest:
test       25756
control    24244
Name: abtest, dtype: int64
Unique values present in the column gearbox:
manuell      36993
automatik    10327
Name: gearbox, dtype: int64
Unique values present in the column unrepaired_damage:
nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64


After careful observations i decided to drop `seller` and `offer_type` columns. because for `seller` column `49999` of `50000` falls in the categoty `privat`. similarly for `offer_type` column `49999` of `50000` falls in the categoty `Angebot`. 

In [11]:
autos.drop(['seller', 'offer_type'], axis = 1, inplace = True)

In [12]:
autos.columns

Index(['data_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'powerPS', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'no_of_pictures', 'postal_code', 'last_seen'],
      dtype='object')

Now that we've removed the columns, we'll now convert 'odometer' and 'price' to string values.

In [13]:
autos['odometer_km'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer_km, dtype: int64

Remove the `km` from the value and convert the column into numeric type.

In [14]:
autos['odometer_km'] = autos['odometer_km'].str.replace('km','').str.replace(',','').astype(float)

In [15]:
autos['odometer_km'].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

In [16]:
autos['price'].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$414          1
$79,933       1
$5,198        1
$18,890       1
$16,995       1
Name: price, Length: 2357, dtype: int64

In [17]:
# remove the $ sign and convert the column to numeric type
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_6476\3000492017.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)


In [18]:
autos['price'].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
414         1
79933       1
5198        1
18890       1
16995       1
Name: price, Length: 2357, dtype: int64

####  Exploring the Odometer and Price Columns

Next, we'll dive deeper into the series `odometer_km` and `price`. Our goal is to check the columns of data for any outliers. Ultimimately, we want to identify any extreme values and remove them from our data set. First, we'll analyze the `price` series.


In [19]:
autos['price'].unique().shape # output number of unique values in the series

(2357,)

In [20]:
np.round(autos['price'].describe())

count       50000.0
mean         9840.0
std        481104.0
min             0.0
25%          1100.0
50%          2950.0
75%          7200.0
max      99999999.0
Name: price, dtype: float64

In [21]:
top_thirty_prices = autos['price'].value_counts().sort_index(ascending = False).head(30)
bottom_thirty_prices = autos['price'].value_counts().sort_index(ascending = False).tail(30)

In [22]:
top_thirty_prices

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price, dtype: int64

In [23]:
bottom_thirty_prices

66       1
65       5
60       9
59       1
55       2
50      49
49       4
47       1
45       4
40       6
35       1
30       7
29       1
25       5
20       4
18       1
17       3
15       2
14       1
13       2
12       3
11       2
10       7
9        1
8        1
5        2
3        1
2        3
1      156
0     1421
Name: price, dtype: int64

There are 1421 free cars and 156 one dollar cars. Because those two price points account for approximately 3 percent of our total price points, and the bottom 25 percent of our standard distribution is 1,100 dollars, these two price points don't account for many of the total data points in the set, but are still capable of skewing the entire set. Therefore, we will be considering these two price points outlier totals. On the other hand, we have some very expensive cars. There are 11 cars that cost over 1,000,000 dollars that are in the dataset and 53 over 100,000 dollars. The top 75 percent value in the standard deviation is 7,200 dollars and the mean value is 9,840 dollars. This suggests that these high values, which are only representative of 64 out of the 50,000 cars in the dataset, are skewing the data so much that the average price point is over 7,000 dollars above our median price point. Therefore, we will designate everything over 100,000 dollars as an outlier value.

In [24]:
autos = autos[autos['price'].between(2,100000)]

In [25]:
autos['price'].value_counts().sort_index(ascending = False)

99900    2
99000    2
98500    1
94999    1
93911    1
        ..
9        1
8        1
5        2
3        1
2        3
Name: price, Length: 2308, dtype: int64

In [26]:
autos['odometer_km'].describe()

count     48370.000000
mean     125856.109159
std       39662.917522
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [27]:
top_five_km = autos['odometer_km'].value_counts().sort_index(ascending = False).head(5)
bottom_five_km = autos['odometer_km'].value_counts().sort_index(ascending = False).tail(5)

In [28]:
top_five_km

150000.0    31304
125000.0     5044
100000.0     2106
90000.0      1733
80000.0      1413
Name: odometer_km, dtype: int64

In [29]:
bottom_five_km

40000.0    814
30000.0    773
20000.0    754
10000.0    247
5000.0     806
Name: odometer_km, dtype: int64

Of our remaining datapoints, over 60 percent of the cars sold have 150,000 km on the odometer. because of this, the median, 75 percent total, and max total are all 150,000. But because this accounts for 60 percent of all data points in the entire dataset, I do not feel that it is advisable to throw out over half of the rows of data. Therefore, I will not be removing any outliers from this series. Plus, the cars in the dataset are used, and it is very common for a used car to be sold with many miles on it.


#### Exploring the date columns

There are a number of columns that represent date values:

- date_crawled
- last_seen
- ad_created
- registration_month
- registration_year

In the DataFrame, `date_crawled`, `last_seen` and `ad_created` are columns that are all represented by strings. These will need to be converted to a numeric representation we can analyze. The `registration_month` and `registration_year` columns are already represented as numeric values.

First, let's see how the values in these three string columns are formatted.

In [30]:
autos[['data_crawled', 'last_seen', 'ad_created', 'registration_month',
       'registration_year']].head()

data_crawled            last_seen           ad_created  \
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00   
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00   
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00   
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00   
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

`data_crawled`, `last_seen`, and `ad_created` columns have `time` information along the date, but we are only interested in the date. so we are going to extract only the date part. that is extract the first 10 characters.

In [31]:
autos[['data_crawled', 'last_seen', 'ad_created', 'registration_month',
       'registration_year']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48370 entries, 0 to 49999
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   data_crawled        48370 non-null  object
 1   last_seen           48370 non-null  object
 2   ad_created          48370 non-null  object
 3   registration_month  48370 non-null  int64 
 4   registration_year   48370 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 2.2+ MB


In [32]:
# Include the missing values in the distribution
# rank the date in ascending order

(autos['data_crawled'].str[:10]
                      .value_counts(normalize = True, dropna = False)
                      .sort_index())

2016-03-05    0.025346
2016-03-06    0.014079
2016-03-07    0.036076
2016-03-08    0.033264
2016-03-09    0.033058
2016-03-10    0.032231
2016-03-11    0.032624
2016-03-12    0.036965
2016-03-13    0.015671
2016-03-14    0.036655
2016-03-15    0.034257
2016-03-16    0.029522
2016-03-17    0.031631
2016-03-18    0.012880
2016-03-19    0.034774
2016-03-20    0.037813
2016-03-21    0.037296
2016-03-22    0.032851
2016-03-23    0.032272
2016-03-24    0.029398
2016-03-25    0.031569
2016-03-26    0.032231
2016-03-27    0.031114
2016-03-28    0.034836
2016-03-29    0.034153
2016-03-30    0.033699
2016-03-31    0.031817
2016-04-01    0.033699
2016-04-02    0.035518
2016-04-03    0.038578
2016-04-04    0.036510
2016-04-05    0.013066
2016-04-06    0.003163
2016-04-07    0.001385
Name: data_crawled, dtype: float64

- When `normalize` is set to `True`, the resulting counts are normalized to represent relative frequencies instead of absolute counts. This means that the values are expressed as percentages of the total count.(normalize=True will give you the percentage distribution of each unique value in that column.)

- When `dropna` is set to `False`, it means that missing values (NaN) will be included in the count.

- Here we can see data has been crawled from march 2016to april 2016, and can see an uniformity in days.

In [33]:
(autos['ad_created'].str[:10]
                    .value_counts(normalize=True, dropna=False)
                    .sort_index(). head(20))

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
Name: ad_created, dtype: float64

Ad has created within the same time frame, but some ads were created in 2015.

In [34]:
(autos['last_seen'].str[:10]
                   .value_counts(normalize=True, dropna=False)
                   .sort_index())

2016-03-05    0.001075
2016-03-06    0.004342
2016-03-07    0.005417
2016-03-08    0.007381
2016-03-09    0.009613
2016-03-10    0.010626
2016-03-11    0.012384
2016-03-12    0.023816
2016-03-13    0.008869
2016-03-14    0.012632
2016-03-15    0.015878
2016-03-16    0.016436
2016-03-17    0.028096
2016-03-18    0.007319
2016-03-19    0.015836
2016-03-20    0.020633
2016-03-21    0.020612
2016-03-22    0.021398
2016-03-23    0.018586
2016-03-24    0.019723
2016-03-25    0.019206
2016-03-26    0.016767
2016-03-27    0.015609
2016-03-28    0.020881
2016-03-29    0.022349
2016-03-30    0.024767
2016-03-31    0.023858
2016-04-01    0.022886
2016-04-02    0.024850
2016-04-03    0.025202
2016-04-04    0.024519
2016-04-05    0.124809
2016-04-06    0.221625
2016-04-07    0.132003
Name: last_seen, dtype: float64

#### Dealing with Incorrect Registration Year Data

One thing that stands out from the exploration we did in the last is that the `registration_year` column contains some odd values:

- The minimum value is 1000, before cars were invented
- The maximum value is 9999, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s. Let's count the number of listings with cars that fall outside the 1900 - 2016 interval 


In [35]:
round(autos['registration_year'].between(1900, 2016, inclusive = True).sum() / autos.shape[0] * 100, 2)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_6476\2227843449.py:1: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated in favour of `both` or `neither`.
  round(autos['registration_year'].between(1900, 2016, inclusive = True).sum() / autos.shape[0] * 100, 2)


96.12

`96.12%` `registration_year` column values are between the range `1900` to `2016`. only `4%` of values in the column are above this range, so we can drop those values above or below `1900` and `2016`

In [36]:
autos = autos[(autos['registration_year'] >= 1900) & (autos['registration_year'] <= 2016)]
autos['registration_year'].describe()

count    46492.000000
mean      2002.924008
std          7.150978
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

Now the registration year column is more accurate. years are between the range 1900 and 2016

#### Exploring Price by Brand

In [37]:
top_brands = list(autos['brand'].value_counts()[:6].index)
top_brands

['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford']

Above are the top five brands in the dataframe, we are going to calulate the mean price of each brand

In [38]:
brands_mean_price = {}

for brand in top_brands:
    brands_mean_price[brand] = round(autos[autos['brand'] == brand]['price'].mean(), 3)

In [39]:
brands_mean_price

{'volkswagen': 5417.79,
 'bmw': 8235.22,
 'opel': 2990.72,
 'mercedes_benz': 8558.283,
 'audi': 9321.058,
 'ford': 3718.744}

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between

#### Storing Aggregate Data in a DataFrame

we created `brands_mean_price` next step is to convert this dictionary into a pandas Series

In [40]:
autos.columns

Index(['data_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'powerPS', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'no_of_pictures', 'postal_code', 'last_seen'],
      dtype='object')

In [41]:
autos['odometer_km']
brands_mean_milage = {}

for brand in top_brands:
    brands_mean_milage[brand] = round(autos[autos['brand'] == brand]['odometer_km'].mean(), 3)
    

In [46]:
brands_mean_milage

{'volkswagen': 128708.562,
 'bmw': 132614.903,
 'opel': 129322.458,
 'mercedes_benz': 130901.895,
 'audi': 129239.265,
 'ford': 124233.487}

In [47]:
brands_mean_milage

{'volkswagen': 128708.562,
 'bmw': 132614.903,
 'opel': 129322.458,
 'mercedes_benz': 130901.895,
 'audi': 129239.265,
 'ford': 124233.487}

In [50]:
milage = pd.Series(brands_mean_milage)
milage

volkswagen       128708.562
bmw              132614.903
opel             129322.458
mercedes_benz    130901.895
audi             129239.265
ford             124233.487
dtype: float64

In [44]:
price_series = pd.Series(brands_mean_price)
price_series

volkswagen       5417.790
bmw              8235.220
opel             2990.720
mercedes_benz    8558.283
audi             9321.058
ford             3718.744
dtype: float64

The keys in the dictionary became the index in the series object. We can then create a single-column dataframe from this series object. We need to use the columns parameter when calling the dataframe constructor (which accepts an array-like object) to specify the column name (or the column name will be set to 0 by default):

In [57]:
df = pd.DataFrame(price_series, columns = ['mean_price'])
df

mean_price
volkswagen       5417.790
bmw              8235.220
opel             2990.720
mercedes_benz    8558.283
audi             9321.058
ford             3718.744

In [58]:
df['mean_milage'] = milage

In [62]:
df = df.reset_index().rename(columns = {'index' : 'top_brands'})
df

top_brands  mean_price  mean_milage
0     volkswagen    5417.790   128708.562
1            bmw    8235.220   132614.903
2           opel    2990.720   129322.458
3  mercedes_benz    8558.283   130901.895
4           audi    9321.058   129239.265
5           ford    3718.744   124233.487

#### Conclusions
Our analysis of eBay Kleinanzeigen's listings showed the influence of price, mileage, and unrepaired damage on a car's overall value. We identified the most popular cars on the website listings, examined the relationship between mileage and price, then explored the effects of unrepaired damage on a car's value.